File: csharp/src/Zentinelle/Resilience/CircuitBreaker.cs
namespace Zentinelle.Resilience;

/// <summary>
/// Circuit breaker for failing fast when the service is unavailable.
/// </summary>
internal class CircuitBreaker
{
    private readonly int _failureThreshold;
    private readonly TimeSpan _recoveryTimeout;
    private readonly int _halfOpenMaxCalls;
    private readonly object _lock = new();

    private CircuitState _state = CircuitState.Closed;
    private int _failureCount;
    private int _halfOpenCalls;
    private DateTime? _lastFailureTime;

    /// <summary>
    /// Current state of the circuit breaker.
    /// </summary>
    public CircuitState State
    {
        get
        {
            lock (_lock)
            {
                return _state;
            }
        }
    }

    /// <summary>
    /// Creates a new circuit breaker.
    /// </summary>
    public CircuitBreaker(int failureThreshold, TimeSpan recoveryTimeout, int halfOpenMaxCalls = 3)
    {
        _failureThreshold = failureThreshold;
        _recoveryTimeout = recoveryTimeout;
        _halfOpenMaxCalls = halfOpenMaxCalls;
    }

    /// <summary>
    /// Checks if an operation can be executed.
    /// </summary>
    public bool CanExecute()
    {
        lock (_lock)
        {
            switch (_state)
            {
                case CircuitState.Closed:
                    return true;

                case CircuitState.Open:
                    if (_lastFailureTime.HasValue &&
                        DateTime.UtcNow - _lastFailureTime.Value > _recoveryTimeout)
                    {
                        _state = CircuitState.HalfOpen;
                        _halfOpenCalls = 0;
                        return true;
                    }
                    return false;

                case CircuitState.HalfOpen:
                    return true;

                default:
                    return false;
            }
        }
    }

    /// <summary>
    /// Records a successful operation.
    /// </summary>
    public void RecordSuccess()
    {
        lock (_lock)
        {
            switch (_state)
            {
                case CircuitState.HalfOpen:
                    _halfOpenCalls++;
                    if (_halfOpenCalls >= _halfOpenMaxCalls)
                    {
                        _state = CircuitState.Closed;
                        _failureCount = 0;
                    }
                    break;

                case CircuitState.Closed:
                    _failureCount = 0;
                    break;
            }
        }
    }

    /// <summary>
    /// Records a failed operation.
    /// </summary>
    public void RecordFailure()
    {
        lock (_lock)
        {
            _failureCount++;
            _lastFailureTime = DateTime.UtcNow;

            if (_state == CircuitState.HalfOpen)
            {
                _state = CircuitState.Open;
            }
            else if (_failureCount >= _failureThreshold)
            {
                _state = CircuitState.Open;
            }
        }
    }

    /// <summary>
    /// Resets the circuit breaker to closed state.
    /// </summary>
    public void Reset()
    {
        lock (_lock)
        {
            _state = CircuitState.Closed;
            _failureCount = 0;
            _halfOpenCalls = 0;
            _lastFailureTime = null;
        }
    }
}

/// <summary>
/// Circuit breaker states.
/// </summary>
public enum CircuitState
{
    /// <summary>Circuit is closed, operations allowed.</summary>
    Closed,

    /// <summary>Circuit is open, operations blocked.</summary>
    Open,

    /// <summary>Circuit is testing if service recovered.</summary>
    HalfOpen
}

File: java/src/main/java/ai/zentinelle/CircuitBreaker.java
package ai.zentinelle;

import java.time.Duration;
import java.time.Instant;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Circuit breaker for failing fast when service is down.
 *
 * Uses a lock-based approach for correct half-open state semantics:
 * - Tracks successful calls in half-open state (not just allowed calls)
 * - Prevents race conditions between concurrent requests
 */
class CircuitBreaker {

    private enum State { CLOSED, OPEN, HALF_OPEN }

    private final int failureThreshold;
    private final Duration recoveryTimeout;
    private final int halfOpenMaxCalls;

    private State state = State.CLOSED;
    private int failureCount = 0;
    private int halfOpenSuccesses = 0;  // Track successful calls, not just allowed
    private final AtomicInteger halfOpenInFlight = new AtomicInteger(0);  // Currently executing calls
    private Instant lastFailureTime;
    private final Object lock = new Object();

    CircuitBreaker(int failureThreshold, Duration recoveryTimeout) {
        this.failureThreshold = failureThreshold;
        this.recoveryTimeout = recoveryTimeout;
        this.halfOpenMaxCalls = 3;
    }

    boolean canExecute() {
        synchronized (lock) {
            switch (state) {
                case CLOSED:
                    return true;

                case OPEN:
                    if (lastFailureTime != null &&
                        Duration.between(lastFailureTime, Instant.now()).compareTo(recoveryTimeout) > 0) {
                        // Transition to half-open
                        state = State.HALF_OPEN;
                        halfOpenSuccesses = 0;
                        halfOpenInFlight.set(0);
                        // Allow this call
                        halfOpenInFlight.incrementAndGet();
                        return true;
                    }
                    return false;

                case HALF_OPEN:
                    // Limit concurrent calls in half-open state
                    if (halfOpenInFlight.get() < halfOpenMaxCalls) {
                        halfOpenInFlight.incrementAndGet();
                        return true;
                    }
                    return false;

                default:
                    return false;
            }
        }
    }

    void recordSuccess() {
        synchronized (lock) {
            switch (state) {
                case HALF_OPEN:
                    halfOpenInFlight.decrementAndGet();
                    halfOpenSuccesses++;
                    // Close circuit after N successful calls
                    if (halfOpenSuccesses >= halfOpenMaxCalls) {
                        state = State.CLOSED;
                        failureCount = 0;
                        halfOpenSuccesses = 0;
                    }
                    break;

                case CLOSED:
                    // Reset failure count on success
                    failureCount = 0;
                    break;

                default:
                    break;
            }
        }
    }

    void recordFailure() {
        synchronized (lock) {
            failureCount++;
            lastFailureTime = Instant.now();

            switch (state) {
                case HALF_OPEN:
                    halfOpenInFlight.decrementAndGet();
                    // Any failure in half-open goes back to open
                    state = State.OPEN;
                    break;

                case CLOSED:
                    if (failureCount >= failureThreshold) {
                        state = State.OPEN;
                    }
                    break;

                default:
                    break;
            }
        }
    }
}
